Tasks/DownloadBuildArtifacts/main.ts

```diff
@@ -0,0 +1,139 @@
+        var parallelLimit : number = +tl.getVariable("release.artifact.download.parallellimit");
+
+        var templatePath: string = path.join(__dirname, 'vsts.handlebars.txt');
+        var buildApi : IBuildApi = webApi.getBuildApi();
+        var artifacts = [];
+        var itemPattern : string = '**';
+
+        // verfiy that buildId belongs to the definition selected
+        if (definitionId) {
+            var builds = await buildApi.getBuilds(projectId, [parseInt(definitionId)]).catch((reason) => {
+                reject(reason);
+            });
+
+            if (builds) {
+                var buildIds = builds.map((value, index) => {
+                    return value.id;
+                });
+
+                if (buildIds.indexOf(buildId) == -1) {
+                    reject(tl.loc("BuildIdBuildDefinitionMismatch", buildId, definitionId));
+                }
+            }
+            else {
+                reject(tl.loc("NoBuildsFound", definitionId));
+            }
+        }
+
+        // populate itempattern and artifacts based on downloadType
+        if (downloadType === 'single') {
+            var artifactName = tl.getInput("artifactName");
+            var artifact = await buildApi.getArtifact(buildId, artifactName, projectId).catch((reason) => {
+                reject(reason);
+            });
+            artifacts.push(artifact);
+            itemPattern = artifactName + '/**';
+        }
+        else {
+            var buildArtifacts = await buildApi.getArtifacts(buildId, projectId).catch((reason) => {
+                reject(reason);
+            });
+
+            console.log(tl.loc("LinkedArtifactCount", buildArtifacts.length));
+            artifacts = artifacts.concat(buildArtifacts);
+            itemPattern = tl.getInput("itemPattern", false) || '**';
+        }
+
+        var downloadPromises: Array<Promise<any>> = [];
+
+        artifacts.forEach(async function (artifact, index, artifacts) {
+            let downloaderOptions = new engine.ArtifactEngineOptions();
+            downloaderOptions.itemPattern = itemPattern;
+            downloaderOptions.verbose = isVerbose;
+
+            if(parallelLimit){
+                downloaderOptions.parallelProcessingLimit = parallelLimit;
+            }
+
+            if (artifact.resource.type.toLowerCase() === "container") {
+                let downloader = new engine.ArtifactEngine();
+                var containerParts: string[] = artifact.resource.data.split('/', 3);
+                if (containerParts.length !== 3) {
+                    throw new Error(tl.loc("FileContainerInvalidArtifactData"));
+                }
+
+                var containerId: number = parseInt(containerParts[1]);
+                var containerPath: string = containerParts[2];
+
+                var itemsUrl = endpointUrl + "/_apis/resources/Containers/" + containerId + "?itemPath=" + containerPath + "&isShallow=true";
```

Comment:
Why aren't you using the file container API?

Reply:
Not sure what you mean, we are calling file container api only. If you are taking about using file container client, then we aren't using it we are using typed-rest-client in item-level-downloader for all apis.

Reply:
Why are you constructing a URL by concatenating strings?  We have strongly-typed clients for a reason.  You are using the build client in this same method.  Why are you not using the file container client?

Reply:
and does this mean that when we have package service artifacts, we'll be manually constructing URLs for those, too?

Reply:
@scottdallamura - Download Build Artifact task can choose to use any strongly typed VSTS HTTP client. But we wouldn't want to take dependency on these clients within item level downloader. 

In case package service artifacts cannot be downloaded using item level downloader, we will let the Download Build Artifacts task directly download them using their VSTS HTTP client.